Implement MAP_UNION_SUM aggregation

kaikalur opened the pull request:
This is a heavily used aggregation that aggregates maps into one by adding values for matching keys. SQL/lambda versions are inefficient. So we implement them natively.

Test plan - Added tests

```
== RELEASE NOTES ==

General Changes
*  :func `map_union_sum`
 
```

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,220 @@
+public class KeySetAndValues
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(KeySetAndValues.class).instanceSize();
+    private static final int EXPECTED_ENTRIES = 10;
```
Call these `DEFAULT_xxx` maybe?

kaikalur replied:
we hace this pattern in other places so just following that

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,220 @@
+                valueType.createBlockBuilder(null, EXPECTED_ENTRIES, expectedValueSize(valueType, EXPECTED_ENTRY_SIZE)));
+    }
+
+    public KeySetAndValues(Type keyType, Type valueType, TypedSet keySet, BlockBuilder valueBlockBuilder)
```
Does this need to be public?

kaikalur replied:
fixed

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,220 @@
+
+    public long estimatedInMemorySize()
+    {
+        long size = INSTANCE_SIZE;
```
Just use `return INSTANCE_SIZE + ... + ...`. We normally don't create locals in this situation.

kaikalur replied:
Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,220 @@
+        }
+    };
+
+    private static interface Summer
```
The name sounds really strange to me. What about "aggregator" or "accumulator"?

kaikalur replied:
I purposely left it here what people will say lol. Yeah I will change it. It's not aggregator or accumulator. It's an adder. I can call it Adder

kaikalur replied:
Done

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,217 @@
+    public KeySetAndValues unionSum(KeySetAndValues other)
+    {
+        Adder adder = getSummer(valueType);
+        TypedSet resultKeySet = new TypedSet(keyType, max(keySet.size(), other.keySet.size()), "MAP_UNION_SUM");
```
Is it necessary to create the `resultKeySet` always? I thought you included the TypedSet into the aggregation state because you wanted to save some cost of building the hash tables. It should work to use the existing `keySet` member and add the elements in `other` to it because this is union semantic and no elements need to be deleted from existing `keySet`. The algorithm is something like this:
```
int[] matchedPositions = new int[valueBlockBuilder.getPositionCount()];
fill matchedPositions with -1;

for each element of position i in other {
    check if keySet contains it. If yes {
        matchedPositions[position in keySet] = i;
    }
    else {
        insert the current element to this.keySet. Note we have calculated the hash position already when testing contains.
        append this element to this.valueBlockBuilder;
     }
}

BlockBuilder resultValueBlockBuilder = valueType.createBlockBuilder(null, EXPECTED_ENTRIES, expectedValueSize(valueType, EXPECTED_ENTRY_SIZE));
for each position i in this.valueBlockBuilder, 
        if (matchedPositions[i] != -1) {
        append the sum (valueBlockBuilder.get(i) + otherValueBlock.get(matchedPositions[i]) to the resultValueBlockBuilder;
    }
    else {
        append the current value in this.valueBlockBuilder to resultValueBlockBuilder
    }
}
return new KeySetAndValues(keyType, valueType, keySet, resultValueBlockBuilder);
```
This way we can save some memory for creating a whole new KeySetAndValues, and should save the cost of adding the elements in `keySet` to `resultKeySet` which requires to compute hash positions again.

kaikalur replied:
Yeah - see the latest code - it's totally different. In fact, I'm refactoring it so we will only one set/group

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,161 @@
+    @Override
+    public String getDescription()
+    {
+        return "Aggregate all the maps into a single map";
```
This description doesn't have the sum logic.

kaikalur replied:
fixed

yingsu00 commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -5546,4 +5546,32 @@ public void testApproxMostFrequentWithStringGroupBy()
+    @Test
+    public void testMapUnionSum()
+    {
+        MaterializedResult actual = computeActual("select map_union_sum(x) from (select map(array['x', 'y'], cast(array[1.1,2] as array<real>) ) x union all select map(array['x', 'y'], cast(array[10,20] as array<real>)))");
```
I see all tests are testing two maps. Would it make sense to test more than 2?

kaikalur replied:
Done

yingsu00 reviewed (commented):
I'm curious: why do not implement this function using similar way to MapConcatFunction, and make the aggregation state just the input map blocks? The code would be simpler, and the state to send/receive is much smaller. A `KeySetAndValues` object contains a TypedSet and the whole BlockBuilder, and inside of TypedSet there is another BlockBuilder and a hash table. I was guessing you include the TypedSet in the state to save some computation, but the implementation didn't fully make use of it.

Also, would it be possible to add a JMH benchmark?

yingsu00 commented:
Just a thought: I would create a `OptimizedTypedMap` that supports more generic "add" operation for this function. Using `TypedSet` is not very efficient.

kaikalur commented:
Latest version eliminates creating new typedsets.

kaikalur commented:
> I'm curious: why do not implement this function using similar way to MapConcatFunction, and make the aggregation state just the input map blocks? The code would be simpler, and the state to send/receive is much smaller. A `KeySetAndValues` object contains a TypedSet and the whole BlockBuilder, and inside of TypedSet there is another BlockBuilder and a hash table. I was guessing you include the TypedSet in the state to save some computation, but the implementation didn't fully make use of it.
> 
> Also, would it be possible to add a JMH benchmark?

Yeah - I'm actually reworking the whole thing to have a single typedset.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,217 @@
+        void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder);
+    }
+
+    private static Adder getSummer(final Type type)
```
Oh looks like I missed that one. will fix it.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,217 @@
+            return INT_ADDER;
+        }
+
+        assert (false);
```
that's too much work lol

prithvip commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,159 @@
+
+        // Check for numeric types.
+        if (!TypeUtils.isNonDecimalNumericType(valueType)) {
+            throw new PrestoException(EXPECTING_NUMERIC_TYPE, String.format("MAP_UNION_SUM expects value type of map to be one of numeric type, but got: %s", valueType.getDisplayName()));
```
Shouldn't this be caught at analysis time?

kaikalur replied:
That's too much work :)

prithvip commented on presto-docs/src/main/sphinx/functions/aggregate.rst:
```diff
@@ -193,6 +193,11 @@ Map Aggregate Functions
 
+.. function:: map_union_sum(x(K,V)) -> map(K,V)
+
+      Returns the union of all the input maps SUM'ing the values of matching keys in all
```
"SUM'ing" should be "summing"? 
And I think in this case, since this is not a standard SQL function, we should add an example illustrating what this function does.

kaikalur replied:
Done

prithvip commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -5546,4 +5546,36 @@ public void testApproxMostFrequentWithStringGroupBy()
     }
+
+    @Test
+    public void testMapUnionSum()
```
Can we add a test that fails - for example on a non-numeric map value?

kaikalur replied:
Done

kaikalur commented:
> Just a thought: I would create a `OptimizedTypedMap` that supports more generic "add" operation for this function. Using `TypedSet` is not very efficient.

Actually, it's not too bad, especially for this operation as we need to keep all the values and do the addition.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,217 @@
+        }
+    };
+
+    private static interface Adder
```
> I'm curious: why do not implement this function using similar way to MapConcatFunction, and make the aggregation state just the input map blocks? The code would be simpler, and the state to send/receive is much smaller. A `KeySetAndValues` object contains a TypedSet and the whole BlockBuilder, and inside of TypedSet there is another BlockBuilder and a hash table. I was guessing you include the TypedSet in the state to save some computation, but the implementation didn't fully make use of it.
> 

Please take a look now. We have only one typedset lazily built

> Also, would it be possible to add a JMH benchmark?

This is the first version. So I will let the benchmark for whoever improves it later.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,217 @@
+        void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder);
+    }
+
+    private static Adder getSummer(final Type type)
```
fixed

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,217 @@
+            int valuePosition = keySet.positionOf(otherKeyBlock, i);
+            int otherValuePosition = i;
+            resultKeySet.add(otherKeyBlock, i);
+            if (remaining > 0 && keySet.contains(otherKeyBlock, i)) {
```
Done

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,217 @@
+        out.closeEntry();
+    }
+
+    public long estimatedInMemorySize()
```
> Rename to `getRetainedSizeInBytes` to match with the codebase convention?

Done

kaikalur replied:
> nit: Move to the end of the public methods section

Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,323 @@
+        this(keyType, valueType, null, null, mapBlock);
+    }
+
+    KeySetAndValues(Type keyType, Type valueType, TypedSet keySet, Block valueBlock, Block mapBlock)
```
private

kaikalur replied:
done

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,161 @@
+
+        // Check for numeric types.
+        if (!TypeUtils.isNonDecimalNumericType(valueType)) {
+            throw new PrestoException(EXPECTING_NUMERIC_TYPE, String.format("MAP_UNION_SUM expects value type of map to be one of numeric type, but got: %s", valueType.getDisplayName()));
```
You actually mean non-decimal numeric type but the error message didn't say so.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,161 @@
+        Type valueType = boundVariables.getTypeVariable("V");
+
+        // Check for numeric types.
+        if (!TypeUtils.isNonDecimalNumericType(valueType)) {
```
If I understand this correctly, you added the type constraint so non-numeric types would fail at analyzing time. However, decimal would fail at run time. Feels like adding a new type constraint only to do a half-ass job is really strange. Either do the type constraint properly, or if it's too much work, remove it and let it fail at specialize.

kaikalur replied:
Done. Added tests as well.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,323 @@
+            return INT_ADDER;
+        }
+
+        assert (false);
```
What about TINYINT and SMALLINT? At least change this to throw proper internal errors. This looks very uncivilized. 😂

kaikalur replied:
the main issue is it's bug-prone to sum tinyint due to overflows and stuff. So I wasn't sure if that's actually worth implementing. But I can add it.

kaikalur replied:
also, we seem to be upconverting everything to integer? That's why I don't think it's even needed.

kaikalur replied:
Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,323 @@
+            return INT_ADDER;
+        }
+
+        assert (false);
```
Also, just a thought. Does it make sense to support boolean? It could be useful...

kaikalur replied:
That will be MAP_UNION_OR. We can open a separate issue for that.

rongrong replied:
It's more like map_union_any_match, so maybe also map_union_all_match 😂

kaikalur replied:
> As synced offline, please add a description to `KeySetAndValues` to describe how it's supposed to work.

Done - refactored so the logic is lot easier to understand in the two different classes - one for the singlemap case and the other for combine.

rongrong reviewed (changes requested):
Please change the commit message conforming to guideline. Also the release notes.

kaikalur commented on presto-docs/src/main/sphinx/functions/aggregate.rst:
```diff
@@ -193,6 +193,11 @@ Map Aggregate Functions
+.. function:: map_union_sum(x(K,V)) -> map(K,V)
+
+      Returns the union of all the input maps summing the values of matching keys in all
+      the maps. All null values in the original maps are coalesced to 0.
```
It's a subtle thing if there is a only one row and has some keys mapping to null, they will be coalesed to 0 as well so 'null values ignored' doesn't sound right.

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,161 @@
+            throw new PrestoException(EXPECTING_NUMERIC_TYPE, String.format("MAP_UNION_SUM expects value type of map to be one of numeric type, but got: %s", valueType.getDisplayName()));
+        }
+
+        MapType outputType = (MapType) functionAndTypeManager.getParameterizedType(StandardTypes.MAP, ImmutableList.of(
```
Done

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,161 @@
+    {
+        KeySetAndValues pairs = state.get();
+
+        if (pairs == null) {
```
Done

kaikalur commented:
Update: I finally figured out why my type constraint is not working. So cleaned up that part, added the correct type constraint, removed my check and added tests. I also fixed the addition to use the type-specific add operator for better consistency.

kaikalur commented:
> Please change the commit message conforming to guideline. Also the release notes.

Done

kaikalur commented:
> Please change the commit message conforming to guideline. Also the release notes.

I think I did.

rongrong commented on presto-common/src/main/java/com/facebook/presto/common/type/TypeUtils.java:
```diff
@@ -35,6 +41,28 @@ private TypeUtils()
     {
     }
 
+    public static boolean isNumericType(Type type)
```
Since you are using composition for `isNonDecimalNumericType`, should this be changed to `isNonDecimalNumericType(type) || type instance DecimalType` then?

kaikalur replied:
Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,311 @@
+
+    public void serialize(BlockBuilder out)
+    {
+        assert mapBlock != null != (keySet != null && valueBlock != null);
```
This assert is too fancy. Can we just do `mapBlock != null && ...`

kaikalur replied:
Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java:
```diff
@@ -0,0 +1,311 @@
+        // Just create the keyset
+        TypedSet resultKeySet = this.keySet;
+        if (resultKeySet == null) {
+            resultKeySet = new TypedSet(keyType, mapBlock.getPositionCount() / 2, "MAP_UNION_SUM");
```
This means that both keySets are null right? When can this happen?

kaikalur replied:
Adding two single states. It should be clearer now with the two different classes. Just trying to do it in the class that already has a keyset.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,151 @@
+                new ParameterMetadata(INPUT_CHANNEL, inputType));
+    }
+
+    public static void input(Type keyType, Type valueType, MapUnionSumState state, Block value)
```
Can we change the `Block value` to `Block input`? value and valueBlock/valueType is totally different which makes it very confusing.

kaikalur replied:
Done

rongrong replied:
not done...

kaikalur replied:
Done for real this time :)

rongrong reviewed (commented):
As synced offline, please add a description to `KeySetAndValues` to describe how it's supposed to work.

kaikalur reviewed (commented):
Refactored it to simplify the logic into single map result vs combining maps.

kaikalur commented:
> Please change the commit message conforming to guideline. Also the release notes.

Done.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java:
```diff
@@ -65,6 +65,7 @@
     INVALID_ANALYZE_PROPERTY(0x0000_002A, USER_ERROR),
     GENERATED_BYTECODE_TOO_LARGE(0x0000_002B, USER_ERROR),
     WARNING_AS_ERROR(0x0000_002C, USER_ERROR),
+    EXPECTING_NUMERIC_TYPE(0x0000_002D, USER_ERROR),
```
Is this used?

kaikalur replied:
Removed

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumResult.java:
```diff
@@ -0,0 +1,321 @@
+    {
+        if (other instanceof KeySetAndValues &&
+                (this instanceof SingleMapBlock || size() < other.size())) {
+            // We can avoid create a new set by merging into the other one.
```
why would merging into the other one avoid creating a new set?

kaikalur replied:
The typed set continues to grow and we don't need to create a new one.

rongrong replied:
"avoid creating a new key set"?

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumAggregation.java:
```diff
@@ -0,0 +1,151 @@
+    public static final MapUnionSumAggregation MAP_UNION_SUM = new MapUnionSumAggregation();
+    public static final String NAME = "map_union_sum";
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(MapUnionSumAggregation.class, "output", MapUnionSumState.class, BlockBuilder.class);
+    private static final MethodHandle INPUT_FUNCTION = methodHandle(MapUnionSumAggregation.class, "input", Type.class, Type.class, MapUnionSumState.class, Block.class);
```
Done

kaikalur commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:
```diff
@@ -157,6 +157,11 @@ public int positionOf(Block block, int position)
         return blockPositionByHash.get(getHashPositionOfElement(block, position));
     }
 
+    public BlockBuilder getBlock()
```
Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumResult.java:
```diff
@@ -0,0 +1,321 @@
+            return other.unionSum(this);
+        }
+
+        TypedSet resultKeySet = buildKeySet();
```
Seems like this function don't always "build" a key set, maybe name it `getKeySet`?

kaikalur replied:
done
